common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+    std::shared_ptr<Socket> m_socket;
+};
+
+class ClientSocketAbstractImpl : public ClientSocket {
+public:
+    // TODO: add connection timeout
+    std::unique_ptr<Connection> connect(const SockAddr &address) override
+    {
+        if (0 != ::connect(m_socket->fd(), address.sockaddr(), address.length())) {
+            return nullptr;
+        }
+        return std::make_unique<SocketConnectionImpl>(m_socket);
+    }
+
+protected:
+    explicit ClientSocketAbstractImpl(const std::shared_ptr<Socket> &socket) : m_socket(socket) {}
+    std::shared_ptr<Socket> m_socket;
+};
+
+class RawServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit RawServerSocket(const std::shared_ptr<RawSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class RawClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit RawClientSocket(const std::shared_ptr<RawSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdpServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit UdpServerSocket(const std::shared_ptr<UdpSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdpClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit UdpClientSocket(const std::shared_ptr<UdpSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class TcpServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit TcpServerSocket(const std::shared_ptr<TcpSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class TcpClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit TcpClientSocket(const std::shared_ptr<TcpSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdsServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit UdsServerSocket(const std::shared_ptr<UdsSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdsClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit UdsClientSocket(const std::shared_ptr<UdsSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+/**
+ * prplmesh-democracy: Do you prefer templates?
+ * These template classes would replace previous ServerSocket and ClientSocket derived classes
+ */
+template <class SocketType> class ServerSocketImpl : public ServerSocketAbstractImpl {
+public:
+    explicit ServerSocketImpl(const std::shared_ptr<SocketType> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+template <class SocketType> class ClientSocketImpl : public ClientSocketAbstractImpl {
+public:
+    explicit ClientSocketImpl(const std::shared_ptr<SocketType> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+} // namespace sockets
+
+/**
+ * A CMDU handler class is the equivalent in functionality to, for example, each of those methods

```

Comment:
AFAIU this is pretty much completely independent of the socket stuff above, right? To the extent that we can even introduce it into the code in two separate phases?

Reply:
Yes, they're completely independent (indeed classes from here on are out of the `sockets` namespace).